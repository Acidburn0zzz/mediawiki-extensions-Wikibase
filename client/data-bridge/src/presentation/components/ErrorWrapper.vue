<template>
	<section class="wb-db-error">
		<ErrorPermission
			v-if="permissionErrors.length"
			:permission-errors="permissionErrors"
		/>
		<ErrorUnsupportedDatatype
			v-else-if="unsupportedDatatypeError !== null"
			:data-type="unsupportedDatatypeError.info.unsupportedDatatype"
		/>
		<ErrorDeprecatedStatement
			v-else-if="statementValueIsDeprecated"
		/>
		<ErrorUnknown v-else />
	</section>
</template>

<script lang="ts">
import Component, { mixins } from 'vue-class-component';
import { MissingPermissionsError, PageNotEditable } from '@/definitions/data-access/BridgePermissionsRepository';
import StateMixin from '@/presentation/StateMixin';
import ErrorPermission from '@/presentation/components/ErrorPermission.vue';
import ErrorUnknown from '@/presentation/components/ErrorUnknown.vue';
import ErrorUnsupportedDatatype from '@/presentation/components/ErrorUnsupportedDatatype.vue';
import ErrorDeprecatedStatement from '@/presentation/components/ErrorDeprecatedStatement.vue';
import ApplicationError, {
	ErrorTypes,
	UnsupportedDatatypeError,
} from '@/definitions/ApplicationError';

@Component( {
	components: {
		ErrorPermission,
		ErrorUnknown,
		ErrorUnsupportedDatatype,
		ErrorDeprecatedStatement,
	},
} )
export default class ErrorWrapper extends mixins( StateMixin ) {
	public get applicationErrors(): ApplicationError[] {
		return this.rootModule.state.applicationErrors;
	}

	public get permissionErrors(): MissingPermissionsError[] {
		return this.applicationErrors.filter( this.isPermissionError );
	}

	private isPermissionError( error: ApplicationError ): error is MissingPermissionsError {
		return ( Object.values( PageNotEditable ) as string[] ).includes( error.type );
	}

	public get unsupportedDatatypeError(): UnsupportedDatatypeError|null {
		for ( const applicationError of this.applicationErrors ) {
			if ( applicationError.type === ErrorTypes.UNSUPPORTED_DATATYPE ) {
				return applicationError;
			}
		}
		return null;
	}

	public get statementValueIsDeprecated(): boolean {
		return this.applicationErrors.some(
			( applicationError ) => applicationError.type === ErrorTypes.UNSUPPORTED_DEPRECATED_STATEMENT,
		);
	}
}
</script>

<style lang="scss">
.wb-db-error {
	padding: $padding-panel-form;
}
</style>
