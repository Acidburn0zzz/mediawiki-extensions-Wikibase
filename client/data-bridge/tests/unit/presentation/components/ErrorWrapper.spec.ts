import ErrorWrapper from '@/presentation/components/ErrorWrapper.vue';
import { shallowMount } from '@vue/test-utils';
import MediaWikiRouter from '@/definitions/MediaWikiRouter';

describe( 'ErrorWrapper', () => {
	it( 'renders correctly', () => {
		const wrapper = shallowMount( ErrorWrapper );
		expect( wrapper.element ).toMatchSnapshot();
	} );

	it( 'shows the repository about-link', () => {
		const aboutArticleUrl = 'http://localhost/wiki/Project:About';
		const $repoRouter: MediaWikiRouter = {
			getPageUrl: jest.fn().mockReturnValue( aboutArticleUrl ),
		};
		const wrapper = shallowMount( ErrorWrapper, {
			mocks: {
				$repoRouter,
			},
		} );

		expect( $repoRouter.getPageUrl ).toHaveBeenCalledWith( 'Project:About' );
		expect( wrapper.find( 'a' ).attributes( 'href' ) )
			.toBe( aboutArticleUrl );
	} );
} );
