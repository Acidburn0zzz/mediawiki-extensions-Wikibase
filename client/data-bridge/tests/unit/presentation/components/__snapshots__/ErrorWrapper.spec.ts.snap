// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorWrapper renders correctly 1`] = `
<section
  class="wb-db-error"
>
  <h1>
    An error occurred
  </h1>
   
  <p>
    <a
      href="Project:About"
    >
      About the repository
    </a>
    ,
		
    <a
      href="Special:Log/protect"
    >
      An interesting protection history
    </a>
    .
	
  </p>
</section>
`;
